app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,229 @@
+package org.kiwix.kiwixmobile.zim_manager.local_file_transfer;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.IntentFilter;
+import android.net.wifi.WpsInfo;
+import android.net.wifi.p2p.WifiP2pConfig;
+import android.net.wifi.p2p.WifiP2pDevice;
+import android.net.wifi.p2p.WifiP2pDeviceList;
+import android.net.wifi.p2p.WifiP2pInfo;
+import android.net.wifi.p2p.WifiP2pManager;
+import android.os.Build;
+import android.util.Log;
+import android.widget.Toast;
+import androidx.annotation.NonNull;
+import java.net.InetAddress;
+import org.kiwix.kiwixmobile.R;
+
+import static android.os.Looper.getMainLooper;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.showToast;
+
+
+/**
+ * Manager for the Wifi-P2p API, used in the local file transfer module
+ * */
+public class WifiDirectManager implements WifiP2pManager.ChannelListener, WifiP2pManager.PeerListListener, WifiP2pManager.ConnectionInfoListener {
+
+  private static final String TAG = "WifiDirectManager";
+
+  @NonNull LocalFileTransferActivity activity;
+
+  /* Variables related to the WiFi P2P API */
+  private boolean wifiP2pEnabled = false; // Whether WiFi has been enabled or not
+  private boolean retryChannel = false;   // Whether channel has retried connecting previously
+
+  private WifiP2pManager manager;         // Overall manager of Wifi p2p connections for the module
+  private WifiP2pManager.Channel channel;
+  // Connects the module to device's underlying Wifi p2p framework
+
+  private final IntentFilter intentFilter = new IntentFilter();
+  // For specifying broadcasts (of the P2P API) that the module needs to respond to
+  private BroadcastReceiver receiver = null; // For receiving the broadcasts given by above filter
+
+  private WifiP2pDevice userDevice;   // Represents the device on which the app is running
+  private WifiP2pInfo groupInfo;      // Corresponds to P2P group formed between the two devices
+
+  private WifiP2pDevice senderSelectedPeerDevice = null;
+
+  public WifiDirectManager(@NonNull LocalFileTransferActivity activity) {
+    this.activity = activity;
+  }
+
+  /* Initialisations for using the WiFi P2P API */
+  public void initialiseWifiDirectManager() {
+    // Intents that the broadcast receiver will be responding to
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+
+    manager = (WifiP2pManager) activity.getSystemService(Context.WIFI_P2P_SERVICE);
+    channel = manager.initialize(activity, getMainLooper(), null);
+  }
+
+  public void registerWifiDirectBroadcastRecevier() {
+    receiver = new WifiDirectBroadcastReceiver(manager, channel, activity);
```

Comment:
You will probably be able to inject this BroadcastReceiver in the constructor as well, during construction you can register it and then call unregister during onDestroy. You don't want the current behaviour of onPause/onResume because that means if the user leaves the screen (presses the home button, answers a call) you will stop listening to these wifi events

Reply:
You probably can't inject this anymore, I'll resolve this conversation